src/transformers/generation_utils.py

```diff
@@ -1938,43 +1980,18 @@ def contrastive_search(
                         else (outputs.hidden_states,)
                     )
 
-            # enlarge the past_key_values
+            # Replicates the new past_key_values to match the `top_k` candidates
             new_key_values = []
-            for layer in past_key_values:
+            for layer in model_kwargs["past"]:
                 items = []
                 # item is either the key or the value matrix
                 for item in layer:
-                    bsz, num_head, seq_len, esz = item.size()
-                    item = (
-                        item.unsqueeze(1)
-                        .expand(-1, top_k, -1, -1, -1)
-                        .reshape(bsz * top_k, num_head, seq_len, esz)
-                        .contiguous()
-                    )  # [bsz*beam, num_head, seq_len, esz]
-                    items.append(item)
+                    items.append(item.repeat_interleave(top_k, dim=0))
                 new_key_values.append(items)
-            past_key_values = new_key_values
-
-            # build next attention mask
-            if "attention_mask" in model_inputs:
-                attention_mask = model_kwargs["attention_mask"]  # [B, S]
-                attention_mask = attention_mask.unsqueeze(1).expand(-1, top_k, -1).reshape(-1, attention_mask.size(-1))
-            else:
-                attention_mask = None
+            model_kwargs["past"] = new_key_values
 
-            # encoder-decoder model also contains the `encoder_outputs`
-            if self.config.is_encoder_decoder and "encoder_outputs" in model_inputs:
-                encoder_outputs = model_inputs["encoder_outputs"]
-            else:
-                encoder_outputs = None
-            next_model_inputs = self.prepare_inputs_for_generation(
-                top_k_ids.view(-1, 1),
-                past=past_key_values,
-                attention_mask=attention_mask,
-                use_cache=True,
-                encoder_outputs=encoder_outputs,
-            )
             # compute the candidate tokens by the language model and collects their hidden_states
+            next_model_inputs = self.prepare_inputs_for_generation(top_k_ids.view(-1, 1), **model_kwargs)
```

Comment:
With the standardization of operations in the past two PRs, `model_kwargs` holds all the information needed here (just like in the other generate methods)